File: rustprogramming/stack_queue/20_valid_parentheses.rs
struct Solution;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = vec![];
        for ch in s.chars() {
            if stack.is_empty() {
                stack.push(ch);
            } else {
                let last = *stack.last().unwrap();
                match (last, ch) {
                    ('(', ')') | ('[', ']') | ('{', '}') => {stack.pop();},
                    _ => stack.push(ch),
                };
            }
        }
        return stack.is_empty();
    }
}

fn main() {
    println!("{}", Solution::is_valid(String::from("()")));
    println!("{}", Solution::is_valid(String::from("()[]{}")));
    println!("{}", Solution::is_valid(String::from("(]")));
    println!("{}", Solution::is_valid(String::from("([)]")));
    println!("{}", Solution::is_valid(String::from("{[]}"))); 
}

File: codingpuzzles/20_valid_parentheses.cc
#include <map>
#include <stack>
#include <iostream>

using namespace std;

class Solution {
public:
    // bool isValid(string s) {
    //     stack<char> _stack;
    //     for (char& ch : s) {
    //         if (_stack.size() == 0) {
    //             _stack.push(ch);
    //         } else {
    //             if (_stack.top() == '(' && ch == ')' || _stack.top() == '[' && ch == ']' || _stack.top() == '{' && ch == '}') {
    //                 _stack.pop();
    //             } else {
    //                 _stack.push(ch);
    //             }
    //         }
    //     }
    //     return _stack.size() == 0;
    // }

    bool isValid(string s) {
        stack<char> _stack;
        map<char, char> _map {{')', '('}, {']', '['}, {'}', '{'}};
        
        for (char& ch : s) {
            if (_stack.size() == 0 || _map.find(ch) == _map.end()) {
                _stack.push(ch);
            } else {
                if (_map[ch] == _stack.top()) {
                    _stack.pop();
                } else {
                    return false;
                }
            }
        }
        return _stack.size() == 0;
    }
};

int main() {
    printf("%d", Solution().isValid("()"));
    printf("%d", Solution().isValid("()[]{}"));
    printf("%d", Solution().isValid("(]"));
    printf("%d", Solution().isValid("([)]"));
    printf("%d", Solution().isValid("{[]}"));
    return 0;
}